src/operator/tensor/la_op_inline.h

```diff
@@ -0,0 +1,335 @@
+  template<typename xpu, typename DType>
+  static void op(const Tensor<xpu, 2, DType>& dA, const Tensor<xpu, 2, DType>& L,
+                 const Tensor<xpu, 2, DType>& A, const Tensor<xpu, 2, DType>& dL,
+                 const nnvm::NodeAttrs& attrs) {
+    // Backward of A = potri(L).
+    // dL = -2 * tril(A * dA * L**(-T)), where tril() extracts lower triangle and diagonal.
+    gemm::op(A, dA, dL, DType(1.0), DType(0), false, false);
+    trsm::op(L, dL, DType(-2.0), true, true);
+    ZeroUpper(dL.dptr_, dL.size(0));
+  }
+};
+
+struct trsm_backward {
+  template<typename xpu, typename DType>
+  static void op(const Tensor<xpu, 2, DType>& dB, const Tensor<xpu, 2, DType>& L,
+                 const Tensor<xpu, 2, DType>& A, const Tensor<xpu, 2, DType>& B,
+                 const Tensor<xpu, 2, DType>& dL, const Tensor<xpu, 2, DType>& dA,
+                 const nnvm::NodeAttrs& attrs) {
+    // Backward of B = trsm(L,A).
+    const LaTriangMatrixMultParam& param = nnvm::get<LaTriangMatrixMultParam>(attrs.parsed);
+    // Compute dA
+    if ( dA.dptr_ != dB.dptr_ ) Copy(dA, dB);
+    trsm::op(L, dA, DType(param.alpha), param.rightside, !param.transpose);
+    // Compute dL
+    const bool da_left(param.rightside == param.transpose);
+    (da_left ?
+        gemm::op(dA, B, dL, DType(-1.0/param.alpha), DType(0), param.transpose, !param.transpose)
+      : gemm::op(B, dA, dL, DType(-1.0/param.alpha), DType(0), !param.transpose, param.transpose));
+    ZeroUpper(dL.dptr_, dL.size(0));
+  }
+};
+
+struct trmm_backward {
+  template<typename xpu, typename DType>
+  static void op(const Tensor<xpu, 2, DType>& dB, const Tensor<xpu, 2, DType>& L,
+                 const Tensor<xpu, 2, DType>& A, const Tensor<xpu, 2, DType>& B,
+                 const Tensor<xpu, 2, DType>& dL, const Tensor<xpu, 2, DType>& dA,
+                 const nnvm::NodeAttrs& attrs) {
+    // Backward of B = trmm(L,A).
+    const LaTriangMatrixMultParam& param = nnvm::get<LaTriangMatrixMultParam>(attrs.parsed);
+    // Compute dL
+    const bool db_left(param.rightside == param.transpose);
+    (db_left ? gemm::op(dB, A, dL, DType(param.alpha), DType(0), param.transpose, !param.transpose)
+           : gemm::op(A, dB, dL, DType(param.alpha), DType(0), !param.transpose, param.transpose));
+    ZeroUpper(dL.dptr_, dL.size(0));
+    // Compute dA
+    if ( dA.dptr_ != dB.dptr_ ) Copy(dA, dB);
+    trmm::op(L, dA, DType(param.alpha), param.rightside, !param.transpose);
+  }
+};
+
+struct sumlogdiag_backward {
+  template<typename xpu, typename DType>
+  static void op(const DType& dB, const Tensor<xpu, 2, DType>& A, const Tensor<xpu, 2, DType>& dA,
+                 const nnvm::NodeAttrs& attrs, bool add) {
+    // Backward of B = sumlogdiag(A).
+    const int N(A.size(0));
+    if ( !add ) {
+      for ( int i = 0; i < N*N; ++i ) {
+        dA.dptr_[i] = 0;
+      }
+    }
+    for ( int i = 0; i < N; ++i ) {
+      dA.dptr_[i*(N+1)] += dB / A.dptr_[i*N+i];
```

Comment:
Certainly. Though these linear algebra operators comprise a package for advanced numerical applications. For such applications, numerical stability is a big issue in general, and usually is the callers responsibility. At the lower level, there is no way to ensure numerical stability anymore. For example the matrices that are valid input to potrf must be symmetric and positive semidefinite and this may not be the case  because of some rounding problems in prior steps of the algorithm. It is the algorithm designers responsibility to ensure that proper mechanisms are added such that everything is numerically stable. Dividing by zero is only one special case here, so no special mechanisms are added for this case.